src/python/pants/base/specs.py

```diff
@@ -1,94 +1,84 @@
 # Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+import itertools
 import os
-import re
 from abc import ABC, ABCMeta, abstractmethod
 from dataclasses import dataclass
-from typing import Any, Dict, Iterable, Iterator, List, Optional, Sequence, Tuple, Union, cast
+from typing import TYPE_CHECKING, Iterable, Mapping, Optional, Sequence, Tuple, Union, cast
 
+from pants.base.exceptions import ResolveError
+from pants.build_graph.address import Address
 from pants.engine.collection import Collection
 from pants.engine.fs import GlobExpansionConjunction, GlobMatchErrorBehavior, PathGlobs
+from pants.engine.internals.target_adaptor import TargetAdaptor
 from pants.util.collections import assert_single_element
 from pants.util.dirutil import fast_relpath_optional, recursive_dirname
-from pants.util.filtering import and_filters, create_filters
 from pants.util.memo import memoized_property
-from pants.util.meta import frozen_after_init
+
+if TYPE_CHECKING:
+    from pants.engine.internals.mapper import AddressFamily
 
 
 class Spec(ABC):
     """A specification for what Pants should operate on."""
 
     @abstractmethod
-    def to_spec_string(self) -> str:
-        """Return the normalized string representation of this spec."""
+    def __str__(self) -> str:
+        """The normalized string representation of this spec."""
```

Comment:
Possibly this change wasn't much of an improvement. I originally had a `.spec` property to mirror `Address`, but realized in Python that implementing `__str__` is more idiomatic.